ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -47,6 +48,359 @@ struct PendingRenameOperation
+                    }
+                }
+                RWLock_ReleaseShared(KextLogTracer::traceFilterLock);
+            }
+            else
+            {
+                snprintf(pathBuffer, sizeof(pathBuffer), "[vn_getpath() failed: %d]", error);
+            }
+        }
+        
+        return shouldTrace;
+    }
+    
+    uint64_t GetTraceIndex()
+    {
+        if (!this->hasTraceIndex)
+        {
+            this->traceIndex = atomic_fetch_add(&KextLogTracer::nextTraceIndex, 1llu);
+            this->hasTraceIndex = true;
+        }
+        return this->traceIndex;
+    }
+    
+    TraceBuffer GetBuffer()
+    {
+        if (this->dynamicTraceBuffer != nullptr)
+        {
+            return TraceBuffer { this->dynamicTraceBuffer, this->dynamicTraceBufferSize, this->traceBufferPosition };
+        }
+        else
+        {
+            return TraceBuffer { this->embeddedTraceBuffer, sizeof(this->embeddedTraceBuffer), this->traceBufferPosition };
+        }
+    }
+    
+    bool GrowBuffer(uint32_t minimumSize)
+    {
+        TraceBuffer currentBuffer = this->GetBuffer();
+        if (minimumSize <= currentBuffer.size)
+        {
+            return true;
+        }
+        
+        uint32_t newSize = MAX(minimumSize, currentBuffer.size * 2u);
+        char* newBuffer = static_cast<char*>(Memory_Alloc(newSize));
+        if (newBuffer == nullptr)
+        {
+            return false;
+        }
+        
+        memcpy(newBuffer, currentBuffer.buffer, currentBuffer.position);
+        newBuffer[currentBuffer.position] = '\0';
+        
+        if (this->dynamicTraceBuffer != nullptr)
+        {
+            Memory_Free(this->dynamicTraceBuffer, this->dynamicTraceBufferSize);
+        }
+        this->dynamicTraceBuffer = newBuffer;
+        this->dynamicTraceBufferSize = newSize;
+        
+        return true;
+    }
+
+    void Emit()
+    {
+        if (!this->discarded)
+        {
+            this->willEmitTrace = true;
+
+            TraceBuffer currentBuffer = this->GetBuffer();
+            uint64_t index = this->GetTraceIndex();
+            KextLog("Event trace %9llu: %.*s", index, currentBuffer.position, currentBuffer.buffer);
+        }
+    }
+
+    void Flush()
+    {
+        if (!this->discarded)
+        {
+            this->Emit();
+            TraceBuffer currentBuffer = this->GetBuffer();
+            currentBuffer.buffer[0] = '\0';
+            this->traceBufferPosition = 0;
+        }
+    };
+public:
+    KextLogTracer(kauth_action_t vnodeAction, vnode_t vnode) :
+        traceBufferPosition(0),
+        dynamicTraceBuffer(nullptr),
+        dynamicTraceBufferSize(0),
+        discarded(false),
+        willEmitTrace(false)
+    {
+        this->embeddedTraceBuffer[0] = '\0';
+        
+        kauth_action_t mask = atomic_load(&KextLogTracer::traceVnodeActionFilterMask);
+        if (0 == (mask & vnodeAction))
```

Comment:
To be clear we drop actions that don't include the mask, but if a mask was never set we accept all?

Reply:
When tracing is first enabled in `PrjFSService::setProperties()`, the [mask defaults to `~0`](https://github.com/microsoft/VFSForGit/pull/1457/files/57466cbe5cb8001cfd1c9d27b701af40521429e2#diff-1beba556f03bf2aee251088144262bc0R150) so all actions allowed. Only if the user process sets it to something else do we deviate from that default.